cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,461 @@
+	}
+	defer func() {
+		if err := opusDec.Destroy(); err != nil {
+			slog.Error("processLiveCaptionsForTrack: failed to destroy decoder", slog.String("err", err.Error()),
+				slog.String("trackID", ctx.trackID))
+		}
+	}()
+
+	// Setup the VAD
+	sd, err := speech.NewDetector(speech.DetectorConfig{
+		ModelPath:  filepath.Join(getModelsDir(), "silero_vad.onnx"),
+		SampleRate: trackOutAudioRate,
+
+		// set WindowSize to 512 to get as fine-grained detection as possible (for when
+		// the number of samples don't cleanly divide into the WindowSize
+		WindowSize:           vadWindowSizeInSamples,
+		Threshold:            vadThreshold,
+		MinSilenceDurationMs: vadMinSilenceDurationMs,
+		SpeechPadMs:          vadSpeechPadMs,
+	})
+	if err != nil {
+		slog.Error("processLiveCaptionsForTrack: failed to create speech detector",
+			slog.String("err", err.Error()))
+	}
+	defer func() {
+		if err := sd.Destroy(); err != nil {
+			slog.Error("processLiveCaptionsForTrack: failed to destroy speech detector", slog.String("err", err.Error()))
+		}
+		slog.Debug("processLiveCaptionsForTrack: finished processing live captions",
+			slog.String("trackID", ctx.trackID))
+	}()
+
+	windowPressureLimitSamples := windowPressureLimitSec * 1000 * trackOutAudioSamplesPerMs
+	window := make([]float32, 0, windowPressureLimitSamples)
+	pcmBuf := make([]float32, trackOutFrameSize)
+
+	// readTrackPktPayloads drains the pktPayload channel (data from the track) and converts it to PCM.
+	readTrackPktPayloads := func() error {
+		for {
+			select {
+			case payload, ok := <-pktPayloads:
+				if !ok {
+					// Exit on channel close
+					return errors.New("closed")
+				}
+				n, err := opusDec.Decode(payload, pcmBuf)
+				if err != nil {
+					slog.Error("failed to decode audio data for live captions",
+						slog.String("err", err.Error()),
+						slog.String("trackID", ctx.trackID))
+				}
+				window = append(window, pcmBuf[:n]...)
+			default:
+				// Done draining
+				return nil
+			}
+		}
+	}
+
+	prevTranscribedPos := 0
+	prevWindowLen := 0
+	var prevAudioAt time.Time
+
+	ticker := time.NewTicker(tickRate)
+	defer ticker.Stop()
+
+	// Algorithm summary:
+	// - Get a cleaned version of the voice (with zeroes where no voice is detected)
+	// - And a list of segments of contiguous speech or silence
+	// - Don't transcribe if data hasn't increased.
+	// - Don't transcribe if new (un-transcribed) data is silence.
+	// - Send the cleaned data (the whole window) to the transcriber pool
+	// - If window goes over its limit, we drop the oldest segments until it's below the limit
+	// - Wait for the transcription (let `tick`s pass so that we're only
+	//   transcribing a particular track once at a time)
+	// - Send the transcription to the plugin to be redistributed to clients.
+	// - finish and wait for next `tick`
+
+	for {
+		select {
+		case <-doneCh:
+			return
+		case <-ticker.C:
+			// empty the waiting pktPayloads
+			if err := readTrackPktPayloads(); err != nil {
```

Comment:
Not in love with the interface here. If we are reading we should be returning something. Keeping it as pure side effect can be slightly confusing.

Reply:
Agreed, fixed. Fixed the `removeShortSpeeches` too so we're always returning the result.